### PR TITLE
Build and push all tests to AWS.

### PR DESCRIPTION
**Pull request by mberhault**
- add DIR and STRIPPED args to testbuild rule
- build all tests and tar them up

We still build sql.test by itself to make the nightly logictest easier
(it can grab that single binary as opposed to grabbing the tarball and
extracting the needed one).

Gzipping the tarball to reduce size. This is particularly critical when
running from the office.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3459)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -79,11 +81,15 @@ install:
...
 	for p in $(TESTS); do \
 	  NAME=$$(basename "$$p"); \
 	  OUT="$$NAME.test"; \
-	  DIR=$$($(GO) list -f {{.Dir}} -tags '$(TAGS)' $$p); \
-	  $(GO) test $(GOFLAGS) -tags '$(TAGS)' -o "$$DIR"/"$$OUT" -ldflags '$(LDFLAGS)' "$$p" $(TESTFLAGS) || exit 1; \
+	  DIRROOT=$${ROOT-$$($(go) list -f {{.Root}} -tags '$(TAGS)' $$p)}; \
+	  OUTDIR=$$($(GO) list -f $${DIRROOT}/{{.ImportPath}} -tags '$(TAGS)' $$p); \
+	  $(GO) test $(GOFLAGS) -tags '$(TAGS)' -o "$$OUTDIR"/"$$OUT" -ldflags '$(LDFLAGS)' "$$p" $(TESTFLAGS) || exit 1; \
+	  # Silence output from strip, it complains when running on packages without tests. \
+	  if [ $${STRIPPED-0} -eq 1 ]; then strip -S "$$OUTDIR"/"$$OUT" >/dev/null 2>&1 || true; fi \
```
Comment (tamird) [inline Makefile]: won't this still send stderr to stdout? I would expect `... 2>$1 > /dev/null`

File: build/build-static-binaries.sh [hunk 2]
```diff
@@ -12,17 +12,26 @@ source $(dirname $0)/build-common.sh
...
     check_static acceptance/acceptance.test
+    check_static ${test_build_dir}/github.com/cockroachdb/cockroach/sql/sql.test
 
     strip -S cockroach
-    strip -S sql/sql.test
-    strip -S acceptance/acceptance.test
+
+    rm -f static-tests.tar.gz
+    # Skip the project/repo part of the path inside the tarball.
+    # Even for stripped binaries, gzip results in 167MB (21s) vs 512MB (3s).
```
Comment (tamird) [inline build/build-static-binaries.sh]: > 167MB (21s) vs 512MB (3s)

Is there a typo here?

Comment (mberhault) [inline build/build-static-binaries.sh]: sorry, this isn't particularly clear. this `tarball size (time spent building)` with gzip and without respectively.

File: Makefile (hunk 1 above)
Comment (mberhault) [inline Makefile]: nope. the order matters here.
`2>&1 >/dev/null`: stderr is redirected to /dev/stdout, but it doesn't follow the later /dev/stdout redirect
`>/dev/null 2>&1`: stdout is redirected first, so the new value of stderr is /dev/null as well

Comment (mberhault) [inline Makefile]: or to put it another way: `1` and `2` are file descriptors pointing to `/dev/{stdout,stderr}`, redirects are executed in order and dereference pointers.

File: run.sh [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+echo foo >&2
```
Comment (bdarnell) [inline run.sh]: Remove this

File: Makefile [hunk 4]
```diff
@@ -79,11 +81,15 @@ install:
 testbuild: TESTS := $(shell $(GO) list -tags '$(TAGS)' $(PKG))
 testbuild: GOFLAGS += -c
 testbuild:
+	ROOT="$(DIR)"; \
 	for p in $(TESTS); do \
 	  NAME=$$(basename "$$p"); \
 	  OUT="$$NAME.test"; \
-	  DIR=$$($(GO) list -f {{.Dir}} -tags '$(TAGS)' $$p); \
-	  $(GO) test $(GOFLAGS) -tags '$(TAGS)' -o "$$DIR"/"$$OUT" -ldflags '$(LDFLAGS)' "$$p" $(TESTFLAGS) || exit 1; \
+	  DIRROOT=$${ROOT-$$($(go) list -f {{.Root}} -tags '$(TAGS)' $$p)}; \
```
Comment (bdarnell) [inline Makefile]: This double-escaping is hard to follow. Time to move this out of the makefile into a .sh script?

What is the lowercase `$(go)` in this line?

File: run.sh (hunk 3 above)
Comment (mberhault) [inline run.sh]: oops.

**Comment by bdarnell** [comment]
LGTM

File: Makefile [hunk 5]
```diff
@@ -79,11 +81,15 @@ install:
 testbuild: TESTS := $(shell $(GO) list -tags '$(TAGS)' $(PKG))
 testbuild: GOFLAGS += -c
 testbuild:
+	ROOT="$(DIR)"; \
 	for p in $(TESTS); do \
```
Comment (tschottdorf) [inline Makefile]: might be time to remove that loop. Then you can also lose the `go list` hack and make this target a little less ugly.

Comment (tschottdorf) [inline Makefile]: I don't think that at this point the binaries need to be put in their package. Putting them into the root should be fine.

File: Makefile (hunk 1 above)
Comment (tschottdorf) [inline Makefile]: why do we eat the exit code here? If I want a stripped binary and something goes wrong, I'd rather know.

File: Makefile (hunk 4 above)
Comment (mberhault) [inline Makefile]: it is a bit tedious, but having an external script makes passing all the various make options hard.
What I could do is keep `testbuild` simple. eg: no stripping, no custom output dir, maybe even a single package, and instead have separate script which does all test binaries stripped in a custom directories. the only parameter would be that directory. However, any changes would have to be synced between the two.
An in-between solution would be to keep testbuild simple and have a testbuildstatic with separate defaults.

And fixed the lowercase go, that was a typo. worked though, surprisingly.

File: build/build-static-binaries.sh [hunk 6]
```diff
@@ -12,17 +12,26 @@ source $(dirname $0)/build-common.sh
...
 # container.
 if [ "${1-}" = "docker" ]; then
+    test_build_dir=$(mktemp -d test-binaries.XXXX)
     time make STATIC=1 build
-    time make STATIC=1 testbuild PKG=./sql
-    time make STATIC=1 testbuild PKG=./acceptance TAGS=acceptance
+    # sql/sql.test is built standalone as well to simplify the nightly logictest runs.
+    time make STATIC=1 testbuild STRIPPED=1 PKG=./sql
+    time make STATIC=1 testbuild STRIPPED=1 PKG=./acceptance TAGS=acceptance
+    time make STATIC=1 testbuild STRIPPED=1 DIR=${test_build_dir}
```
Comment (tschottdorf) [inline build/build-static-binaries.sh]: let's keep the `Makefile` clean. There's no reason to have `DIR` leak into it with my above suggestions afaict.

File: Makefile (hunk 5 above)
Comment (tschottdorf) [inline Makefile]: (the loop can go to your script below).

Comment (mberhault) [inline Makefile]: I was doing that originally but decided against it because:
* if the last component of two packages is the same, we'll overwrite one
* if you see the output of `util.test`, it's helpful to know that the full package name is `store/engine/util`

File: Makefile (hunk 1 above)
Comment (mberhault) [inline Makefile]: As the comment says, strip will fail on non-existing files (when running on a package that does not contain tests).

File: Makefile (hunk 5 above)
Comment (mberhault) [inline Makefile]: I think we do need the loop. if I run `testbuild`, I expect all binaries to be built, just like `test` will run everything by default

**Comment by mberhault** [comment]
Reverted `testbuild` and added `testbuildall` which requires `DIR` and automatically strips

File: Makefile [hunk 7]
```diff
@@ -86,6 +86,23 @@ testbuild:
 	  $(GO) test $(GOFLAGS) -tags '$(TAGS)' -o "$$DIR"/"$$OUT" -ldflags '$(LDFLAGS)' "$$p" $(TESTFLAGS) || exit 1; \
 	done
 
+# Build all tests into DIR and strips each.
+# DIR is required.
+.PHONY: testbuildall
+testbuildall: TESTS := $(shell $(GO) list $(PKG))
+testbuildall: GOFLAGS += -c
+testbuildall:
+ifndef DIR
```
Comment (tschottdorf) [inline Makefile]: alternative heckling against `DIR`:

`find . -type f -name '*.test' -exec mv '{}' "${DIR}{}" ';'` in the script will do the same thing.

File: Makefile (hunk 5 above)
Comment (tschottdorf) [inline Makefile]: I don't think so, `go test -c ./...` also doesn't work. That's the reason for this hack here. Better to just lose it.

**Comment by tschottdorf** [comment]
still against keeping the loop in the `Makefile` and still don't think there's a good reason for `DIR`, but LGTM if you a pronounced opinion otherwise.

File: Makefile (hunk 5 above)
Comment (mberhault) [inline Makefile]: I know, this is why we need the loops, otherwise we have to do a loop outside the makefile. It makes more sense to have the default make rule build everything

File: Makefile (hunk 7 above)
Comment (mberhault) [inline Makefile]: I also don't want to rely on test binaries being the only ones named `*.test`. The second someone adds one of those, we'll have a non-binary in the tarball and cause problems.

File: build/build-static-binaries.sh [hunk 8]
```diff
@@ -12,17 +12,28 @@ source $(dirname $0)/build-common.sh
...
 
     strip -S cockroach
     strip -S sql/sql.test
     strip -S acceptance/acceptance.test
+
+    rm -f static-tests.tar.gz
+    # Skip the project/repo part of the path inside the tarball.
+    # Even for stripped binaries, gzip results in 167MB (21s spent) vs 512MB (3s spent).
+    # It makes a big difference when fetching from outside AWS.
+    time tar cfz static-tests.tar.gz -C ${test_build_dir}/github.com/cockroachdb/ cockroach/
```
Comment (tschottdorf) [inline build/build-static-binaries.sh]: there's also http://imoverclocked.blogspot.com/2015/12/for-love-of-bits-stop-using-gzip.html
